test/gtest/core/test_mc_reduce.h

```diff
@@ -112,256 +66,83 @@ struct ReductionTest<UCC_DT_BFLOAT16, op> {
-            ucc_mc_alloc(&buf2_d_mc_header, n * n_bytes, mtype);
-            buf2_d = (typename T::type *)buf2_d_mc_header->addr;
-            ucc_mc_memcpy(res_d, res_h, n_bytes, mtype, UCC_MEMORY_TYPE_HOST);
-            ucc_mc_memcpy(buf1_d, buf1_h, n_bytes, mtype, UCC_MEMORY_TYPE_HOST);
-            ucc_mc_memcpy(buf2_d, buf2_h, n * n_bytes, mtype,
-                          UCC_MEMORY_TYPE_HOST);
-        }
-
-        return UCC_OK;
-    }
-
-    ucc_status_t free_bufs(ucc_memory_type_t mtype)
-    {
-        if (buf1_h != nullptr) {
-            ucc_mc_free(buf1_h_mc_header);
-        }
-        if (buf2_h != nullptr) {
-            ucc_mc_free(buf2_h_mc_header);
-        }
-        if (res_h != nullptr) {
-            ucc_mc_free(res_h_mc_header);
-        }
-        if (buf1_d != nullptr) {
-            ucc_mc_free(buf1_d_mc_header);
-        }
-        if (buf2_d != nullptr) {
-            ucc_mc_free(buf2_d_mc_header);
-        }
-        if (res_d != nullptr) {
-            ucc_mc_free(res_d_mc_header);
+#define DECLARE_OP_(_op, _UCC_OP, _OP)                          \
+    template<typename T>                                        \
+    class _op {                                                 \
+    public:                                                     \
+    const static ucc_reduction_op_t redop = UCC_OP_ ## _UCC_OP; \
+    T operator()(T arg1, T arg2) {                              \
+        return DO_OP_ ## _OP(arg1, arg2);                       \
+    }                                                           \
+    };                                                          \
+
+#define DECLARE_OP(_op, _OP) DECLARE_OP_(_op, _OP, _OP)
+
+DECLARE_OP(sum, SUM);
+DECLARE_OP(prod, PROD);
+DECLARE_OP(min, MIN);
+DECLARE_OP(max, MAX);
+DECLARE_OP(land, LAND);
+DECLARE_OP(lor, LOR);
+DECLARE_OP(lxor, LXOR);
+DECLARE_OP(band, BAND);
+DECLARE_OP(bor, BOR);
+DECLARE_OP(bxor, BXOR);
+DECLARE_OP_(avg, AVG, SUM);
+
+#define FLOAT_OP_PAIRS(_TYPE) TypeOpPair<UCC_DT_ ## _TYPE, sum>,    \
+        TypeOpPair<UCC_DT_ ## _TYPE, prod>,                         \
+        TypeOpPair<UCC_DT_ ## _TYPE, min>,                          \
+        TypeOpPair<UCC_DT_ ## _TYPE, max>
+
+/* TypeOp pairs that MC Cuda supports */
+#define CUDA_OP_PAIRS TypeOpPair<UCC_DT_INT16, lor>,    \
+        TypeOpPair<UCC_DT_INT16, sum>,                  \
+        TypeOpPair<UCC_DT_INT32, prod>,                 \
+        TypeOpPair<UCC_DT_INT64, bxor>,                 \
+        TypeOpPair<UCC_DT_FLOAT32, avg>,                \
+        FLOAT_OP_PAIRS(INT32),                          \
+        FLOAT_OP_PAIRS(FLOAT32),                        \
+        FLOAT_OP_PAIRS(FLOAT64),                        \
+        FLOAT_OP_PAIRS(BFLOAT16)
+
+using CollReduceTypeOpsCuda = ::testing::Types<CUDA_OP_PAIRS>;
+
+/* Host supports all combos, so add more to tests */
+using CollReduceTypeOpsHost = ::testing::Types<CUDA_OP_PAIRS,
```

Comment:
some pairs are missing compared to original version e.g. UCC_DT_INT16 max

Reply:
I know. We never had ALL the combinations (to limit the time spent in collectives with reductions). I changed it that way: we now test ALL possible combos with test_mc_reduce, AND we test  "every possible dtype" and "every possible op" in the reduce coll tests (reduce, allreduce, reduce_scatter), but now every possible combination of them. 
Also i added explicitly INT32, FLOAT32, FLOAT64 with all arithmetic ops since those are most commonly used. If you think we should add more plz comment - i will extend.